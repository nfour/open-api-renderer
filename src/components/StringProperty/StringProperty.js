import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

export default class StringProperty extends PureComponent {
  render() {
    const { constraints } = this.props;

    if (!constraints) {
      return null;
    }

    const { pattern, minLength, maxLength } = constraints;
    const validations = [];

    if (pattern) {
      validations.push(`/${pattern}/`);
    }

    if (maxLength !== undefined && minLength !== undefined) {
      // Be succint if the minLength is the same maxLength
      // ie. value can only be of `x` length.
      if (maxLength === minLength) {
        validations.push(`${minLength} chars`);
      } else {
        validations.push(`${minLength}-${maxLength} chars`);
      }
    } else if (minLength !== undefined) {
      validations.push(`at least ${minLength} chars`);
    } else if (maxLength !== undefined) {
      validations.push(`at most ${maxLength} chars`);
    }

    if (!validations.length) {
      return null;
    }

    return (
      <span>
        {validations.map(constraint =>
          <span key={constraint} className="string-constraints">{constraint}</span>
        )}
      </span>
    );
  }
}

StringProperty.propTypes = {
  constraints: PropTypes.shape({
    pattern: PropTypes.string,
    maxLength: PropTypes.number,
    minLength: PropTypes.number
  })
};
