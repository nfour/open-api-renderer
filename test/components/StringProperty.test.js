import React from 'react';
import StringProperty from './../../src/components/StringProperty/StringProperty';
import renderer from 'react-test-renderer';

describe('<StringProperty />', () => {
  it('renders nothing if there are no applicable constraints', () => {
    let tree = renderer.create(
      <StringProperty />
    ).toJSON();

    expect(tree).toBeNull();

    tree = renderer.create(
      <StringProperty constraints={{minItems: 1}}/>
    ).toJSON();

    expect(tree).toBeNull();
  });

  it('renders pattern', () => {
    const tree = renderer.create(
      <StringProperty constraints={{pattern: '^[a-zA-Z0-9_-]+$'}} />
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders minLength', () => {
    const tree = renderer.create(
      <StringProperty constraints={{minLength: 2}} />
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders maxLength', () => {
    const tree = renderer.create(
      <StringProperty constraints={{maxLength: 10}} />
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders minLength and maxLength', () => {
    const tree = renderer.create(
      <StringProperty constraints={{minLength: 4, maxLength: 10}} />
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders minLength and maxLength when they are the same value', () => {
    const tree = renderer.create(
      <StringProperty constraints={{minLength: 4, maxLength: 4}} />
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders pattern, minLength and maxLength', () => {
    const tree = renderer.create(
      <StringProperty constraints={{pattern: '^[a-z]', minLength: 4, maxLength: 10}} />
    );

    expect(tree).toMatchSnapshot();
  });
});
