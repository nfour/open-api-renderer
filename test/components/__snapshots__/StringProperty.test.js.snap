// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StringProperty /> renders maxLength 1`] = `
<span>
  <span
    className="string-constraints"
  >
    at most 10 chars
  </span>
</span>
`;

exports[`<StringProperty /> renders minLength 1`] = `
<span>
  <span
    className="string-constraints"
  >
    at least 2 chars
  </span>
</span>
`;

exports[`<StringProperty /> renders minLength and maxLength 1`] = `
<span>
  <span
    className="string-constraints"
  >
    4-10 chars
  </span>
</span>
`;

exports[`<StringProperty /> renders minLength and maxLength when they are the same value 1`] = `
<span>
  <span
    className="string-constraints"
  >
    4 chars
  </span>
</span>
`;

exports[`<StringProperty /> renders pattern 1`] = `
<span>
  <span
    className="string-constraints"
  >
    /^[a-zA-Z0-9_-]+$/
  </span>
</span>
`;

exports[`<StringProperty /> renders pattern, minLength and maxLength 1`] = `
<span>
  <span
    className="string-constraints"
  >
    /^[a-z]/
  </span>
  <span
    className="string-constraints"
  >
    4-10 chars
  </span>
</span>
`;
